### PR TITLE
`MPRester` lazily get `endpoint` and `api_key`  

### LINKED ISSUE #935: `MP_API_KEY` is retrieved at import time
`MP_API_KEY` is retrieved at import time, perhaps considering getting at `init` time instead https://github.com/janosh/pymatviz/pull/207#issuecomment-2379211936? https://github.com/materialsproject/api/blob/2c89d4c015940daa5c28b20d4a5d757b8ab6b29f/mp_api/client/mprester.py#L72

If someone want to set the environment variable at runtime (assume no `MP_API_KEY` env var is set beforehand), it would fail:
```python
import os
from mp_api.client import MPRester  # >>> os.environ.get("MP_API_KEY", None) being execuated


os.environ["MP_API_KEY"] = "KEY"

mp_api = os.environ.get("MP_API_KEY")
assert mp_api is not None  # >>> passes

struct = MPRester().get_structure_by_material_id("mp-19017")  # >>> fails as MP_API_KEY is None
```

### PR DESCRIPTION
**Pull request by DanielYang59**
### Summary

- Lazily get `endpoint` and `api_key`, to close #935
- [x] Add more unit tests

### CODE REVIEW

File: mp_api/client/mprester.py
Comment (DanielYang59) [inline mp_api/client/mprester.py:63]: Need confirm on this deletion, I didn't see this warning being accessed.

File: tests/test_mprester.py
Comment (DanielYang59) [inline tests/test_mprester.py:36]: Need confirm: would this test suite be intended to run on other CI env other than GitHub (GitLab or some self-hosted runner?)

File: tests/test_mprester.py
Comment (DanielYang59) [inline tests/test_mprester.py:39]: I believe this skip mark should be removed now, as the `MP_API_KEY` setter is put inside the fixture

File: tests/test_mprester.py
Comment (DanielYang59) [inline tests/test_mprester.py:38]: Need confirm: would this test suite be intended to run on other CI env other than GitHub (GitLab or some self-hosted runner?)

**Comment by DanielYang59** [comment]
@tschaume This one is ready for review, thanks!

**Review by janosh — APPROVED** [review]
thanks @DanielYang59, looks great! esp. those new tests 👍

**Comment by DanielYang59** [comment]
Weird, I cannot reproduce that pre-commit error locally (even with Python 3.11 installed):
```
(python311) yang@Yang-MacBook api % python3 -V
Python 3.11.10
(python311) yang@Yang-MacBook api % pre-commit run --all-files
ruff.....................................................................Passed
black....................................................................Passed
blacken-docs.............................................................Passed
check for case conflicts.................................................Passed
check for broken symlinks............................(no files to check)Skipped
check yaml...............................................................Passed
detect destroyed symlinks................................................Passed
fix end of files.........................................................Passed
mixed line ending........................................................Passed
trim trailing whitespace.................................................Passed
codespell................................................................Passed
```

---

The other test failure is owing to [restricted access to repo secret from a fork](https://github.com/janosh/pymatviz/pull/207#discussion_r1779479931):

> With the exception of GITHUB_TOKEN, secrets are not passed to the runner when a workflow is triggered from a forked repository.

And is passing locally:
```
=================================================== test session starts ====================================================
platform darwin -- Python 3.12.6, pytest-8.3.3, pluggy-1.5.0
rootdir: /Users/yang/developer/api
configfile: pyproject.toml
plugins: typeguard-4.3.0
collected 31 items                                                                                                         

tests/test_client.py ...............................                                                                 [100%]
```

File: mp_api/client/core/client.py
Comment (DanielYang59) [inline mp_api/client/core/client.py:183]: Is this an accidental change or different linter version? This is causing linting error.

File: tests/test_client.py
Comment (DanielYang59) [inline tests/test_client.py:71]: Need confirm on this, I'm not sure if this test is intended to test deprecated method `get_data_by_id` or should we replace it?
```
tests/test_client.py::test_generic_get_methods[PhononRester]
tests/test_client.py::test_generic_get_methods[ProvenanceRester]
tests/test_client.py::test_generic_get_methods[RobocrysRester]
tests/test_client.py::test_generic_get_methods[SimilarityRester]
  /Users/yang/developer/api/tests/test_client.py:82: DeprecationWarning: get_data_by_id is deprecated and will be removed soon. Please use the search method instead.
    doc = rester.get_data_by_id(
```

File: mp_api/client/mprester.py
Comment (tschaume) [reply mp_api/client/mprester.py:63]: Should be fine

File: tests/test_client.py
Comment (tschaume) [reply tests/test_client.py:71]: We probably don't need to test the deprecated method but I'd leave the warnings in for now.

**Review by tschaume — APPROVED** [review]
I think it looks good - I'll try to get it merged and released.

File: tests/test_client.py
Comment (DanielYang59) [inline tests/test_client.py:44]: Thanks for reviewing! 

I don't really think this would work because secrets are not passed to the environment when triggered from a fork. Honestly I don't know any way around this except for running test locally, or wait until this PR get merge and run CI from there.

**Comment by tschaume** [comment]
@DanielYang59 release candidate is out for this 🎉 See https://pypi.org/project/mp-api/0.43.0rc0/

**Comment by DanielYang59** [comment]
Great! Thanks a lot for reviewing!

**Pull request merged**
